pkg/controller/vsphere/reconciler.go

```diff
@@ -507,3 +512,43 @@ func (vm *virtualMachine) getNetworkStatusList(client *vim25.Client) ([]NetworkS
 
 	return networkStatusList, nil
 }
+
+// IgnitionConfig returns a slice of option values that set the given data as
+// the guest's ignition config.
+func IgnitionConfig(data []byte) []types.BaseOptionValue {
+	config := EncodeIgnitionConfig(data)
+
+	if config == "" {
+		return nil
+	}
+
+	return []types.BaseOptionValue{
+		&types.OptionValue{
+			Key:   "guestinfo.ignition.config.data",
+			Value: config,
+		},
+		&types.OptionValue{
+			Key:   "guestinfo.ignition.config.encoding",
+			Value: "base64",
+		},
+	}
+}
+
+// EncodeIgnitionConfig attempts to decode the given data until it looks to be
+// plain-text, then returns a base64 encoded version of that plain-text.
+func EncodeIgnitionConfig(data []byte) string {
+	if len(data) == 0 {
+		return ""
+	}
+
+	for {
+		decoded, err := base64.StdEncoding.DecodeString(string(data))
+		if err != nil {
+			break
```

Comment:
nit: may be log something here

Reply:
Like what? This is expected to break. It breaks when the input looks like plain-text, i.e. can no longer be successfully decoded. And the input is considered secret, so it shouldn't be logged.

Reply:
actually don't we want to return err here?

Reply:
No, we want to break and then re-encode as base64. This function ensures that you can pass either plain-text, or already encoded data (e.g. straight from a k8s secret) to `IgnitionConfig()` and have it do the right thing.